protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,965 @@
+                                    },
+                                    ResponseState::Failed => {
+                                        iter.on_failure(&peer);
+                                    },
+                                }
+                            },
+                            None => {
+                                // The iterator is the first to contact this peer.
+                                self.contacted_peers.insert(
+                                    peer.clone().into_owned(),
+                                    PeerState::new(i),
+                                );
+                                return PeersIterState::Waiting(Some(Cow::Owned(peer.into_owned())));
+                            },
+                        }
+                    }
+                    PeersIterState::WaitingAtCapacity => {
+                        match state {
+                            Some(PeersIterState::Waiting(Some(_))) => {
+                                // [`ClosestDisjointPeersIter::next`] returns immediately once a
+                                // [`ClosestPeersIter`] yielded a peer. Thus this state is
+                                // unreachable.
+                                unreachable!();
+                            },
+                            Some(PeersIterState::Waiting(None)) => {}
+                            Some(PeersIterState::WaitingAtCapacity) => {}
+                            Some(PeersIterState::Finished) => {
+                                // `state` is never set to `Finished`.
+                                unreachable!();
+                            },
+                            None => state = Some(PeersIterState::WaitingAtCapacity),
+                        };
+
+                        break;
+                    }
+                    PeersIterState::Finished => break,
+                }
+            }
+        }
+
+        state.unwrap_or(PeersIterState::Finished)
+    }
+
+    /// Only finish the [`ClosestPeersIter`]s that yielded a peer within
+    /// `peers`.
+    ///
+    /// `peers` representing the peers that either returned or put a record.
+    ///
+    /// See [`crate::query::Query::may_finish`] for details.
+    pub fn may_finish<I>(&mut self, peers: I)
+    where
+        I: IntoIterator<Item = PeerId>
+    {
+        for peer in peers {
+            if let Some(PeerState{ initiated_by, .. }) = self.contacted_peers.get_mut(&peer) {
+                self.iters[*initiated_by].finish();
+            }
+        }
+    }
+
+    /// Immediately transitions the iterator to [`PeersIterState::Finished`].
+    pub fn finish(&mut self) {
+        for iter in &mut self.iters {
+            iter.finish();
+        }
+    }
+
+    /// Checks whether the iterator has finished.
+    pub fn is_finished(&self) -> bool {
+        for iter in &self.iters {
+            if iter.is_finished() {
+                return true;
+            }
+        }
+
+        false
```

Comment:
That doesn't seem right, since this iterator should only be considered finished when all of the inner iterators are finished, right? I.e.
```suggestion
    pub fn is_finished(&self) -> bool {
        self.iters.iter().all(|i| i.is_finished())
    }
```

Reply:
Yes, that is rather embarrassing. Thanks for catching this one.